packages/squiggle-lang/src/rescript/GenericDist/GenericDist.res

```diff
@@ -0,0 +1,251 @@
+      PointSetDist.combineAlgebraically(operation, a, b)
+    )
+
+  let runMonteCarlo = (
+    toSampleSet: toSampleSetFn,
+    operation: GenericDist_Types.Operation.arithmeticOperation,
+    t1: t,
+    t2: t,
+  ) => {
+    let operation = Operation.Algebraic.toFn(operation)
+    E.R.merge(toSampleSet(t1), toSampleSet(t2))->E.R2.fmap(((a, b)) => {
+      Belt.Array.zip(a, b)->E.A2.fmap(((a, b)) => operation(a, b))
+    })
+  }
+
+  //I'm (Ozzie) really just guessing here, very little idea what's best
+  let expectedConvolutionCost: t => int = x =>
+    switch x {
+    | #Symbolic(#Float(_)) => 1
+    | #Symbolic(_) => 1000
+    | #PointSet(Discrete(m)) => m.xyShape->XYShape.T.length
+    | #PointSet(Mixed(_)) => 1000
+    | #PointSet(Continuous(_)) => 1000
+    | _ => 1000
+    }
+
+  let chooseConvolutionOrMonteCarlo = (t2: t, t1: t) =>
+    expectedConvolutionCost(t1) * expectedConvolutionCost(t2) > 10000
+      ? #CalculateWithMonteCarlo
+      : #CalculateWithConvolution
+
+  let run = (
+    t1: t,
+    toPointSet: toPointSetFn,
+    toSampleSet: toSampleSetFn,
+    algebraicOp,
+    t2: t,
+  ): result<t, error> => {
+    switch tryAnalyticalSimplification(algebraicOp, t1, t2) {
+    | Some(Ok(symbolicDist)) => Ok(#Symbolic(symbolicDist))
+    | Some(Error(e)) => Error(Other(e))
+    | None =>
+      switch chooseConvolutionOrMonteCarlo(t1, t2) {
+      | #CalculateWithMonteCarlo =>
+        runMonteCarlo(toSampleSet, algebraicOp, t1, t2)->E.R2.fmap(r => #SampleSet(r))
+      | #CalculateWithConvolution =>
+        runConvolution(toPointSet, algebraicOp, t1, t2)->E.R2.fmap(r => #PointSet(r))
+      }
+    }
+  }
+}
+
+let algebraicCombination = AlgebraicCombination.run
+
+//TODO: Add faster pointwiseCombine fn
+let pointwiseCombination = (t1: t, toPointSet: toPointSetFn, operation, t2: t): result<
+  t,
+  error,
+> => {
+  E.R.merge(toPointSet(t1), toPointSet(t2))
+  ->E.R2.fmap(((t1, t2)) =>
+    PointSetDist.combinePointwise(GenericDist_Types.Operation.arithmeticToFn(operation), t1, t2)
+  )
+  ->E.R2.fmap(r => #PointSet(r))
+}
+
+let pointwiseCombinationFloat = (
+  t: t,
+  toPointSet: toPointSetFn,
+  operation: GenericDist_Types.Operation.arithmeticOperation,
+  f: float,
+): result<t, error> => {
+  switch operation {
+  | #Add | #Subtract => Error(GenericDist_Types.DistributionVerticalShiftIsInvalid)
+  | (#Multiply | #Divide | #Exponentiate | #Log) as operation =>
+    toPointSet(t)->E.R2.fmap(t => {
+      //TODO: Move to PointSet codebase
+      let fn = (secondary, main) => Operation.Scale.toFn(operation, main, secondary)
+      let integralSumCacheFn = Operation.Scale.toIntegralSumCacheFn(operation)
+      let integralCacheFn = Operation.Scale.toIntegralCacheFn(operation)
+      PointSetDist.T.mapY(
+        ~integralSumCacheFn=integralSumCacheFn(f),
+        ~integralCacheFn=integralCacheFn(f),
+        ~fn=fn(f),
+        t,
+      )
+    })
+  }->E.R2.fmap(r => #PointSet(r))
```

Comment:
I thought we established that constructors could be function literals? and that you don't need lambda syntax?

Reply:
do you mean like ``r -> #PointSet``? You could do that; I thought it was a bit harder to read (I tried it), not sure

Reply:
Same line: do you like it this way rather than `let foo = switch operation ... ; foo -> ...`?

Reply:
(I edited my first comment in this thread with what I meant). I mean getting rid of the lambda entirely.

Reply:
Do you mean like this? I guess that might be a bit safer, for future debugging.

<img width="900" alt="image" src="https://user-images.githubusercontent.com/377065/160712540-052cbbbb-e9a5-4ead-bbd5-0e37ce196f31.png">

Reply:
yes, that's what I meant by my second comment.